catalyst-toolbox-lib/src/recovery/tally.rs

```diff
@@ -0,0 +1,465 @@
+    let mut inc_tally = true;
+    for fragment_log in fragment_logs {
+        match fragment_log {
+            Ok(PersistentFragmentLog { fragment, time }) => {
+                let block_date = fragment_log_timestamp_to_blockdate(time, &timeframe, &ledger)
+                    .expect("BlockDates should always be valid for logs timestamps");
+
+                dbg!("Fragment processed {}", fragment.hash());
+                let new_fragment = match &fragment {
+                    fragment @ Fragment::VoteCast(_) => {
+                        if let Ok(new_fragment) =
+                            fragment_replayer.replay(fragment.clone()).map_err(|e| {
+                                eprintln!(
+                                    "Fragment {} couldn't be processed:\n\t {:?}",
+                                    fragment.id(),
+                                    e
+                                );
+                            })
+                        {
+                            if vote_start > block_date || vote_end <= block_date {
+                                eprintln!(
+                                    "Fragment {} skipped because it was out of voting time ({}-{}-{})",
+                                    fragment.id(),
+                                    vote_start,
+                                    block_date,
+                                    vote_end,
+                                );
+                                None
+                            } else {
+                                Some(new_fragment)
+                            }
+                        } else {
+                            failed_fragments.push(fragment.clone());
+                            None
+                        }
+                    }
+                    new_fragment @ Fragment::VoteTally(_) => {
+                        if inc_tally {
+                            ledger = increment_ledger_time_up_to(ledger, vote_end);
+                            inc_tally = false;
+                        }
+                        Some(new_fragment.clone())
+                    }
+                    _ => None,
+                };
+                if let Some(new_fragment) = new_fragment {
+                    ledger = ledger.apply_fragment(&ledger.get_ledger_parameters(), &new_fragment, block_date)
+                        .expect("Should be impossible to fail, since we should be using proper spending counters and signatures");
+                }
+            }
+            Err(e) => {
+                eprintln!("Error deserializing PersistentFragmentLog: {:?}", e);
+            }
+        }
+    }
+    Ok((ledger, failed_fragments))
+}
+
+struct FragmentReplayer {
+    wallets: HashMap<Address, Wallet>,
+    spending_counters: HashMap<Address, Vec<u32>>,
+    voteplans: HashMap<VotePlanId, VotePlan>,
+    old_block0_hash: Hash,
+    fees: LinearFee,
+}
+
+impl FragmentReplayer {
+    const CHECK_RANGE: Range<i32> = -50..50;
```

Comment:
The choice of limits needs to be carefully considered. I'd rather use the number of proposals in all vote plans multiplied by two to be absolutely sure. Coupled with the expectation-centered counter picking as detailed in the comment above, this should not be very slow when we expect the majority of ballots to be legitimately created.

Reply:
Ok, makes sense.

Reply:
I implemented as the number of proposals range up and range down, it should be enough:
`-proposals <- counter -> +proposals`